File: homework4/thiago_airoldi/q3_RunningMedian.py
# Maintain Two Heaps
# O(nlog(n)) Time complexity, where n is the number of elements in the stream
# O(n) Space Complexity
# You are given an array of numbers representing a stream received one by one. Return an array representing the median after each new number is received.


import heapq

def runningMedian(stream):


    # The intution behind this problem is maintaining two heaps.
    # Think of each heap like a left hand and a right hand.
    # The left hand is a max heap which stores values less than the current median.
    # the right hand is a min heap which stores values greater than the current median.
    # Both heap sizes must not differ by more than 1 element, so we might need to move elements around to different heaps to satisfy this.
    # If both heaps have the same size after insertion, the median is the mid point of the top elements from each heap.
    # If the heaps have different sizes after insertion, the median is the top element of whichever heap you just added an element to.
    
    if not stream:
        return []


    left = [] # maxheap, all values <= median
    right = [] # minheap, all values >= median

    res = [] # Stores the median at each point in time of the stream

    median = float('-inf')

    for i in range(len(stream)):

        num = stream[i] # Current number in the stream

        
        if len(left) == len(right): # Both heaps have the same size

            # Check if num is greater than or less than our most recent median
            if num > median:
                # If num > median, it must be put inside the right side heap (the min heap)
                heapq.heappush(right, num)

                median = right[0] # The new median is the value at the top of the heap we just added to
            else:
                # num <= median
                heapq.heappush(left, -1 * num)
                median = -1 * left[0] # We must multiply by -1 since heapq is a minheap, therefore in order to use it as a max heap, we needed to push negatives of
                                      # the real values, so now to extract the real values, we need to multiply the popped values by -1


        elif len(left) > len(right): # Left heap has 1 more element than right heap

            if num > median:
                # Num goes into right heap, and since len(right) < len(left), we do not need to move anything around
                heapq.heappush(right, num)

                # Now since both heaps have the same amount of elements, the median is the midpoint of the sum of the top values from each heap
                median = ((-1 * left[0]) + right[0]) / 2
            
            else:
                # num <= median
                # Num must go to the left heap, however since the left heap is already larger than the right heap, we need to move the left
                # heap's top element to the right heap
                moveElem = heapq.heappop(left)
                moveElem = -1 * moveElem # Revert this value back to its original value since it came from the left side (max heap)

                heapq.heappush(right, moveElem)

                # Now add num to the left heap (max heap)
                heapq.heappush(left, -1 * num)

                # Now since both heaps have the same amount of elements, the median is the midpoint of the sum of the top values from each heap
                median = ((-1 * left[0]) + right[0]) / 2

        elif len(left) < len(right):

            if num > median:
                # Num must go into the right heap, but since len(right) > len(left), we need to move the top element from right heap to the left heap,
                # then we can add num to the right heap
                moveElem = heapq.heappop(right)

                heapq.heappush(left, -1 * moveElem)

                # Now add num to right heap
                heapq.heappush(right, num)

                # Now since both heaps have the same amount of elements, the median is the midpoint of the sum of the top values from each heap
                median = ((-1 * left[0]) + right[0]) / 2

            else:
                # num <= median, so num must go into the left heap
                # since len(left) < len(right), we can add num to the left heap without having to move things around
                heapq.heappush(left, -1 * num)

                # Now since both heaps have the same amount of elements, the median is the midpoint of the sum of the top values from each heap
                median = ((-1 * left[0]) + right[0]) / 2


        
        res.append(median)



    return res




# 18 minutes


# Test Cases

stream1 = [1, 11, 4, 15, 12]
stream2 = []
stream3 = [-3, -2, 0, 2, 3]
stream4 = [3, 1, 8]


print(runningMedian(stream1))
print(runningMedian(stream2))
print(runningMedian(stream3))
print(runningMedian(stream4))

File: homework4/phuc_le/q3_RunningMedian.cpp
/*
    Technique:
    - Maintain two heaps

    - Use two heaps to keep track of the numbers seen so far.
        - Max Heap (left)
           - Stores the smaller half of the numbers.
        - Min Heap (right)
           - Stores the larger half of the numbers.
    - Every element in left <= every element in right.
    - The size difference between the heaps is at most 1.
    - If there is an odd number of elements, left stores one extra element.

    - Insert the new number into the appropriate heap.
    - Rebalance the heaps if their sizes differ by more than one.
    - Compute the median:
        - If both heaps have the same size,
            median = (left.top() + right.top()) / 2.0
        - Else,
          median = left.top()

    Time Complexity: O(log n): n = numbers of elements
    Space Complexity: O(n)

    Time spent: 40 minutes
*/

#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class RunningMedian
{
private:
    // Max heap for smaller half
    priority_queue<int> left;

    // Min heap for larger half
    priority_queue<int, vector<int>, greater<int>> right;

public:
    double addNumber(int num)
    {
        // Insert into appropriate heap
        if (left.empty() || num <= left.top())
            left.push(num);
        else
            right.push(num);

        // Balance heaps
        if (left.size() > right.size() + 1)
        {
            right.push(left.top());
            left.pop();
        }
        else if (right.size() > left.size())
        {
            left.push(right.top());
            right.pop();
        }

        // Compute median
        if (left.size() == right.size())
        {
            return (left.top() + right.top()) / 2.0;
        }

        return left.top();
    }
};

int main()
{
    // Test 1: Example case
    {
        vector<int> nums = {1, 11, 4, 15, 12};

        RunningMedian rm;
        cout << "Test 1: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 2: Single element
    {
        vector<int> nums = {5};

        RunningMedian rm;
        cout << "Test 2: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 3: All same numbers
    {
        vector<int> nums = {7, 7, 7, 7, 7};

        RunningMedian rm;
        cout << "Test 3: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 4: Increasing order
    {
        vector<int> nums = {1, 2, 3, 4, 5};

        RunningMedian rm;
        cout << "Test 4: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 5: Decreasing order
    {
        vector<int> nums = {5, 4, 3, 2, 1};

        RunningMedian rm;
        cout << "Test 5: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 6: Negative numbers
    {
        vector<int> nums = {-5, -1, -10, -3};

        RunningMedian rm;
        cout << "Test 6: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 7: Mix of positive and negative
    {
        vector<int> nums = {-2, 10, -5, 8, 0};

        RunningMedian rm;
        cout << "Test 7: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    // Test 8: Large values
    {
        vector<int> nums = {1000000, 2000000, 3000000};

        RunningMedian rm;
        cout << "Test 8: ";
        for (int x : nums)
            cout << rm.addNumber(x) << " ";
        cout << "\n";
    }

    return 0;
}